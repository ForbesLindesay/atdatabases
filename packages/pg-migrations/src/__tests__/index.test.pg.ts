import {spawnSync} from 'child_process';
import connect, {sql} from '@databases/pg';

// We are using sucrase to compile TypeScript on the fly, and each of these tests
// is starting a fresh node process and creating, then destroying a fresh database
// connection, which makes them feel slow overall, even though each individual test
// isn't actually that slow.
jest.setTimeout(30000);
const db = connect();

const env = {...process.env, FORCE_COLOR: 'false'};

function output(proc: ReturnType<typeof spawnSync>) {
  return [
    `stdout:`,
    proc.stdout?.toString('utf8').replace(/^/gm, '  '),
    `stderr:`,
    proc.stderr?.toString('utf8').replace(/^/gm, '  '),
  ].join('\n');
}

afterAll(async () => {
  await db.dispose();
});

test('clean database', async () => {
  await db.query(sql`
    DROP TABLE IF EXISTS users;
    DROP TABLE IF EXISTS atdatabases_migrations_applied;
    DROP TABLE IF EXISTS atdatabases_migrations_version;
  `);
});

test('help', async () => {
  const proc = spawnSync(
    'node',
    ['--require', 'sucrase/register', 'cli', 'help'],
    {cwd: `${__dirname}/..`, env},
  );
  expect(output(proc)).toMatchSnapshot();
  expect(proc.status).toBe(0);
});

test('help apply', async () => {
  const proc = spawnSync(
    'node',
    ['--require', 'sucrase/register', 'cli', 'help', 'apply'],
    {cwd: `${__dirname}/..`, env},
  );
  expect(output(proc)).toMatchSnapshot();
  expect(proc.status).toBe(0);
});

test('apply - missing directory', async () => {
  const proc = spawnSync(
    'node',
    ['--require', 'sucrase/register', 'cli', 'apply'],
    {cwd: `${__dirname}/..`, env: {...env, CI: 'true'}},
  );
  expect(output(proc)).toMatchSnapshot();
  expect(proc.status).toBe(1);
});

test('apply --dry-run', async () => {
  const proc = spawnSync(
    'node',
    [
      '--require',
      'sucrase/register',
      'cli',
      'apply',
      '--dry-run',
      `-D`,
      `${__dirname}/migrations`,
    ],
    {cwd: `${__dirname}/..`, env: {...env, CI: 'true'}},
  );
  expect(output(proc)).toMatchSnapshot();
  expect(proc.status).toBe(0);
});

test('apply', async () => {
  const proc = spawnSync(
    'node',
    [
      '--require',
      'sucrase/register',
      'cli',
      'apply',
      `-D`,
      `${__dirname}/migrations`,
    ],
    {cwd: `${__dirname}/..`, env: {...env, CI: 'true'}},
  );
  expect(output(proc)).toMatchSnapshot();
  expect(proc.status).toBe(0);
});

test('apply - after already appying', async () => {
  const proc = spawnSync(
    'node',
    [
      '--require',
      'sucrase/register',
      'cli',
      'apply',
      `-D`,
      `${__dirname}/migrations`,
    ],
    {cwd: `${__dirname}/..`, env: {...env, CI: 'true'}},
  );
  expect(output(proc)).toMatchSnapshot();
  expect(proc.status).toBe(0);
});

test('apply --dry-run - after already appying', async () => {
  const proc = spawnSync(
    'node',
    [
      '--require',
      'sucrase/register',
      'cli',
      'apply',
      '--dry-run',
      `-D`,
      `${__dirname}/migrations`,
    ],
    {cwd: `${__dirname}/..`, env: {...env, CI: 'true'}},
  );
  expect(output(proc)).toMatchSnapshot();
  expect(proc.status).toBe(0);
});

test('migrations applied', async () => {
  expect(await db.query(sql`SELECT * FROM users`)).toEqual([
    {
      id: 1,
      name: 'Forbes Lindesay',
    },
    {
      id: 2,
      name: 'Eleanor Brodie',
    },
    {
      id: 3,
      name: 'Carl Tedder',
    },
  ]);
});
