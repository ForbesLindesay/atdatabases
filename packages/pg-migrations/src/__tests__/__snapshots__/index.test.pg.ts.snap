// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply - after already appying 1`] = `
"stdout:
  No migrations required
  
stderr:
  "
`;

exports[`apply - missing directory 1`] = `
"stdout:
  
stderr:
  You must supply a directory path for your migrations. You can supply it as either:
  
   - The migrationsDirectory config value using @databases/pg-config
   - The --directory paramter when calling @databases/pg-migrations
  
  "
`;

exports[`apply --dry-run - after already appying 1`] = `
"stdout:
  No migrations required
  
stderr:
  "
`;

exports[`apply --dry-run 1`] = `
"stdout:
  To update your database, run again without --dry-run
  
stderr:
  This database has not been migrated using @databases/pg-migrations version 2.0.0 before.
  Version updates cannot be automaticaly applied when --dry-run is enabled.
  "
`;

exports[`apply 1`] = `
"stdout:
  Applying 00001-init.ts
  Applied 00001-init.ts
  Applying 00002-update.ts
  Applied 00002-update.ts
  Applying 00003-update2.sql
  Applied 00003-update2.sql
  3 migrations applied
  
stderr:
  "
`;

exports[`help 1`] = `
"stdout:
  Available commands:
  
  pg-migrations apply           - Apply any pending migration
  pg-migrations ignore-error    - Permanently ignore an error for a migration
  pg-migrations mark-applied    - Mark migration as applied (without applying it)
  pg-migrations mark-unapplied  - Mark migration as unapplied (without reverting it)
  pg-migrations restore-from-db - Restore a previously applied migration from the database
  
  For more information on a command, run:
  
  pg-migrations help <COMMAND_NAME>
  
stderr:
  "
`;

exports[`help apply 1`] = `
"stdout:
  Apply Migrations
  
  Usage: pg-migrations apply [options]
  
  Parameters:
  -c --database         A connection string for the database you want to connect to (can also be supplied as the environment variable PG_URL).
  -D --directory        The directory containing migrations (can also be supplied via the \\"migrationsDirectory\\" config option).
     --version-table    A table to store the version of @databases used. Defaults to \\"atdatabases_migrations_version\\".
     --migrations-table A table to store the applied migrations. Defaults to \\"atdatabases_migrations_applied\\".
  -e --ignore-error     Error code to ignore (can use multiple times to ignore multiple errors)
  -d --dry-run          Do actually modify the database or file system.
  
stderr:
  "
`;
